presto-spi/src/main/java/com/facebook/presto/spi/statistics/ExternalPlanStatisticsProvider.java

```diff
@@ -13,17 +13,32 @@
  */
 package com.facebook.presto.spi.statistics;
 
+import com.facebook.presto.common.plan.PlanCanonicalizationStrategy;
 import com.facebook.presto.spi.plan.PlanNode;
 import com.facebook.presto.spi.plan.TableScanNode;
 
+import java.util.Map;
 import java.util.function.Function;
 
 public interface ExternalPlanStatisticsProvider
 {
     String getName();
 
-    PlanStatistics getStats(
-            PlanNode plan,
+    /**
+     * Get historical statistics for given plan nodes.
+     * Depending on `useHashBasedStatistics`, implementation may choose to use plan node hashes supplied,
+     * or do its own statistics prediction by reading the plan.
+     *
+     * TODO: Using PlanNode as map key may be expensive, we can use PlanNodeId as map keys
+     */
+    Map<PlanNode, HistoricalPlanStatistics> getStats(
+            Map<PlanNode, Map<PlanCanonicalizationStrategy, String>> planNodeHashes,
+            boolean useHashBasedStatistics,
             Function<PlanNode, String> planPrinter,
             Function<TableScanNode, TableStatistics> tableStatisticsProvider);
+
+    /**
+     * Given plan hashes and corresponding statistics after a query is run, store them for future retrieval.
+     */
+    void putStats(Map<String, HistoricalPlanStatistics> hashesAndStatistics);
```

Comment:
Not a comment so you can just read and close this one:
This is a good interface. So when a developer reads this interface, he/she would understand the interface is to store a set of history stats based on their hashes. Similar idea can be applied to `getStats` as well. A clean interface would be 
```java
Map<String, HistoricalPlanStatistics> getStats(Set<String>);
```

Of course, if you wanna push the interface design even better, you can create your own class `PlanNodeHash` that wraps around `String`, `byte[]`, or `Slice` so that developers understand better what that is.